src/Rules/RuleBase.ts

```diff
@@ -0,0 +1,51 @@
+import {Example, Options, Rule, RuleType} from '../rules';
+import {Option} from '../option';
+
+export default class RuleBase<TOptions extends Options> {
+  register(): Rule {
+    const rule = new Rule(this.name, this.description, this.type, this.safeApply.bind(this), this.examples, this.options);
+    const classInstance = this.constructor as typeof RuleBase;
+    classInstance.rule = rule;
+
+    return rule;
+  }
+
+    static rule: Rule;
+
+    get name(): string {
+      throw new Error('get name() is not implemented');
+    }
+
+    get description(): string {
+      throw new Error('get description() is not implemented');
+    }
+
+    get type(): RuleType {
+      throw new Error('get type() is not implemented');
+    }
```

Comment:
Is there a reason to make these methods as opposed to letting them be properties? I guess you would have the benefit of getting an error when trying to access info when a `name`, `description` or `type` was not specified.

Reply:
`get name()` is not a method, it is a property getter. So the usage gonna be `const name = new TrailingSpaces().name`

Those properties are defined in the `RuleBase.ts` which is the base class and it kinda forces those properties to be overridden in the child classes. TypeScript doesn't have a concept of abstract methods such as in programming languages like Java or C#. So with this `throw new Error` approach we just give a runtime hint.

Reply:
Thanks for this run down. I am gradually learning the nuances of Typescript, so this is helpful for me to learn.